snapcraft_legacy/plugins/v2/initrd.py

```diff
@@ -0,0 +1,178 @@
+        if self._host_arch != self._target_arch:
+            self._cross_building = True
+        self._ubuntu_series = "focal"
+
+    @overrides
+    def get_build_snaps(self) -> Set[str]:
+        base = self._part_info.base
+
+        build_snaps = {
+            base,
+        }
+        return build_snaps
+
+    @overrides
+    def get_build_packages(self) -> Set[str]:
+        build_packages = {
+            "curl",
+            "fakechroot",
+            "fakeroot",
+        }
+        # consider cross-build option
+        if self._cross_building:
+            build_packages |= {
+                f"libfakechroot:{self._target_arch}",
+                f"libfakeroot:{self._target_arch}",
+            }
+        return build_packages
+
+    @overrides
+    def get_build_environment(self) -> Dict[str, str]:
+        return {
+            "UC_INITRD_ROOT_NAME": "uc-initramfs-build-root",
+            "UC_INITRD_ROOT": "${SNAPCRAFT_PART_SRC}/${UC_INITRD_ROOT_NAME}",
+            "KERNEL_MODULES": "${SNAPCRAFT_STAGE}/modules",
+            "KERNEL_FIRMWARE": "${SNAPCRAFT_STAGE}/firmware",
+            "UBUNTU_SERIES": self._ubuntu_series,
+            "UBUNTU_CORE_BASE": "core20",
+            "CRAFT_ARCH_TRIPLET_BUILD_FOR": "${SNAPCRAFT_ARCH_TRIPLET_BUILD_FOR}",
+            "CRAFT_ARCH_BUILD_FOR": self._target_arch,
+            "CRAFT_ARCH_BUILD_ON": self._host_arch,
+            "CRAFT_STAGE": "${SNAPCRAFT_STAGE}",
+            "CRAFT_PART_SRC": "${SNAPCRAFT_PART_SRC}",
+            "CRAFT_PART_BUILD": "${SNAPCRAFT_PART_BUILD}",
+            "CRAFT_PART_INSTALL": "${SNAPCRAFT_PART_INSTALL}",
+        }
+
+    @property
+    def out_of_source_build(self):
+        """Return whether the plugin performs out-of-source-tree builds."""
+        return True
+
+
+def _get_target_architecture() -> str:
+    # TODO: there is bug in snapcraft and target arch is not reported
+    # correctly. As work around check if we are cross building,
+    # to know what is target arch
+    target_arch = None
+    # pylint: disable=invalid-name
+    for i in range(1, len(sys.argv)):
+        if sys.argv[i].startswith("--target-arch="):
+            target_arch = sys.argv[i].split("=")[1]
+        elif sys.argv[i].startswith("--target-arch"):
+            target_arch = sys.argv[i + 1]
+
+    if target_arch is None:
+        # this is native build, use deb_arch
+        # as for native build it's reported correctly
+        target_arch = ProjectOptions().deb_arch
+
+    return target_arch
+
+    # TODO: finalize initrd_build.sh location
+    @overrides
+    def get_build_commands(self) -> List[str]:
+        logger.info("Getting build commands...")
+        return [
+            f"$SNAP/lib/python3.12/site-packages/snapcraft/parts/plugins/initrd_build.sh initrd-modules={self.options.initrd_modules} initrd-firmware={self.options.initrd_firmware} initrd-addons={self.options.initrd_addons} initrd-overlay={self.options.initrd_overlay}"
+        ]
```

Comment:
This is "legacy" behavior so it impacts earlier `base:` snaps. It should have basically the same semantics as the new (`snapcraft/`) behavior, though.

Specifically, the options should all be the same (where applicable; kernel EFI support in ubuntu-core-initramfs didn't hit until 24.04 for instance).